BUG: handle temporal discontinuities in Neuralynx `.ncs` files

KristijanArmeni opened the pull request:
#### Reference issue
Addresses #12247

#### What does this implement/fix?
Tries do deal with temporal gaps in `.ncs` files. Conceptually, `read_raw_neuralynx()` now checks for any temporal differences between consecutive `neo.Block[0].segments` start/stop times and if any of those are larger than the sampling period, call it a gap (i.e. missing > 1 sample). Then, infer the number of missing samples for each gap based on gap times, do the bookkeeping, generate arrays of 0's for each gap data segment and insert them between respective valid segments.

I can read-in a large dataset locally (and `test_neuralyx.py` passes), but now `raw.plot()` will choke (see more below).

Example behavior on a local dataset where largest discontinuity seems to be 30 msecs (sampling rate = 4kHz).
```python
>>> raw = read_raw_neuralynx(dataset, exclude_fname_patterns=file_patterns, preload=False)
Checking files in /local/neuralynx/dataset/
Ignoring .ncs files:
[...]
N = 418 discontinuous Neo segments detected with delta > 0.00025 sec
(max = 0.030219078063964844 sec, min = 0.00031304359436035156)
```

Note, that this assumes that all segments returned by Neo are valid data. I tried asking about his in neo github, but haven't heard back.

#### Additional information: `raw.plot() breaks`

Essentially, calling .plot with some selection `raw.plot(start=10, duration=2)` will choke with the output below. 

My hunch is the reloading of chunks for plotting somehow interacts with my implementation, where `all_data` array constructed internally from neo segments + times of inferred gaps, won't match the number of samples of the array generated as `block = np.zeros(n_channels, stop-start)`.

Not sure how to best troubleshoot this (I can't share the data I use locally to reproduce and the testing dataset can't be used for this since it's continuous), but hopefully there's something obvious I'm missing?

```python
Traceback (most recent call last):
  File "/home/kriarm/miniconda3/envs/mnedev/lib/python3.11/site-packages/mne_qt_browser/_pg_figure.py", line 2958, in run
    data_chunk, times_chunk = browser._load_data(start, stop)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kriarm/git/mne-python/mne/viz/_figure.py", line 321, in _load_data
    return self.mne.inst[:, start : stop + 2]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/home/kriarm/git/mne-python/mne/io/base.py", line 853, in __getitem__
    return self._getitem(item)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/kriarm/git/mne-python/mne/io/base.py", line 860, in _getitem
    data = self._read_segment(start=start, stop=stop, sel=sel)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<decorator-gen-165>", line 12, in _read_segment
  File "/home/kriarm/git/mne-python/mne/io/base.py", line 478, in _read_segment
    _ReadSegmentFileProtector(self)._read_segment_file(
  File "/home/kriarm/git/mne-python/mne/io/base.py", line 2525, in _read_segment_file
    return self.__raw.__class__._read_segment_file(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kriarm/git/mne-python/mne/io/neuralynx/neuralynx.py", line 349, in _read_segment_file
    block[idx] = all_data  # shape = (n_channels, n_samples)
    ~~~~~^^^^^
ValueError: could not broadcast input array from shape (64,990915) into shape (64,990957)
```

larsoner commented:
> Not sure how to best troubleshoot this (I can't share the data I use locally to reproduce and the testing dataset can't be used for this since it's continuous), but hopefully there's something obvious I'm missing?

First, to cut `raw.plot` out of the equation, doing `raw[:, 2 * int(round(raw.info["sfreq"]))]` or so should also replicate the problem. Or maybe even just a `raw.load_data()` would do it.

> Not sure how to best troubleshoot this (I can't share the data I use locally to reproduce and the testing dataset can't be used for this since it's continuous), but hopefully there's something obvious I'm missing?

I would make sure first that `raw.load_data()` works. This will force you to have a model for the complete set of data, and when something is wrong, it will be clear because you can think of things from sample 0 to sample N-1 of the entire recording. Once that works, you need to map from `start, stop` into that segment properly. The problem could be with the initial construction of the whole signal, or with the mapping of `start, stop` into a particular contiguous segment of the whole signal.

KristijanArmeni commented:
@larsoner Fixed the bug on my end and added commits to detect gaps in Neo segments, infer missing samples, add them as 0's, and annotate as `BAD_ACQ_SKIP`.

### Example gaps
For context, this is a ~40-min recording of story listening.

I cannot comment on whether these look like a typical Neuralynx ring buffer error or not since I have never eye-balled these before. But it seems they can be burst-like, interspersed with a very short (1 or 2 sample) valid samples (this is a 4Khz dataset, so timescale is stretched out to microsecond precision in plots). And these are very short, mostly microsecond-level gaps in the case of my datasets.

![sub-006_gaps](https://github.com/mne-tools/mne-python/assets/14061041/2299d685-80cd-4c2e-b139-34de64fdc4f1)
![sub-006_gaps2](https://github.com/mne-tools/mne-python/assets/14061041/e767e6fd-a48b-45fe-808a-d7da47002734)

### Creating annotations
Right now, I add the all information for annotations in `_raw_extras` and annotations can be created like so:
```python
gap_annot = raw._raw_extras[0]["gap_annotations"]
annot = mne.Annotations(
    onset=gap_annot["onset"],
    duration=gap_annot["duration"],
    description=gap_annot["descriptions"],
)
raw.set_annotations(annot)
```
This requires the user to set annotations. Not sure, if it is preferred that annotations should be added under the hood.

KristijanArmeni commented:
@larsoner `read_raw_neuralynx` now automatically sets annotations if gaps are detected. See in [a32ff97](https://github.com/mne-tools/mne-python/pull/12279/commits/a32ff97c8e0b7b38910cf8928a44854c3695ea06) how its done and if in good order.

Example call and feedback on a local dataset with gaps detected:
```
raw = read_raw_neuralynx("/path/to/local/dataset",
   ...:                  exclude_fname_patterns=glob_patterns,
   ...:                  preload=False)
Checking files in /path/to/local/dataset
Ignoring .ncs files:
CSC2.ncs
CSC3.ncs
... # prints all files
xBLT1.ncs
Checking for temporal discontinuities in Neo data segments.
N = 418 discontinuous Neo segments detected with delta > 0.00025 sec.
(max = 0.030219078063964844 sec, min = 0.00031304359436035156)
Annotating gaps as BAD_ACQ_SKIP.

In [2]: raw.annotations
Out[2]: <Annotations | 418 segments: BAD_ACQ_SKIP (418)>
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,31 +139,133 @@ def __init__(
+        # mark as discontinuous any two segments that have
+        # start/stop delta larger than sampling period (1/sampling_rate)
+        logger.info("Checking for temporal discontinuities in Neo data segments.")
+        delta = 1 / info["sfreq"]
```
Just to be careful about float arithmetic being inexact (if you were using `int` sample numbers you would be safe, but time is float) it might be better to put the boundary at the midpoint of the sample numbers (one sample or two samples later, i.e., 1.5 samples later) like:
```suggestion
        delta = 1.5 / info["sfreq"]
```

KristijanArmeni replied:
Thanks! OK, so if I follow correctly it might be safest to:
1. compute start/stop temporal differences (`float`)
2. express those time differences in `int` samples (e.g. `seg_diffs * sampling_rate)`
3. redefine delta in terms of `int` samples (i.e. not `float` sampling period)
4. and check if any `int(gaps) > int(delta)` (in which case the above correction of delta from `1/sr` to `1.5/sr` is not needed, since we'd be doing exact int ops?)

Happy to do either way, just double-checking on my end!

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,31 +139,133 @@ def __init__(
+            # (n_gaps,) array of ints giving number of samples per inferred gap
+            gap_n_samps = np.array(
+                [
+                    len(np.arange(on_spl, off_spl))
+                    for on_spl, off_spl in zip(
+                        gap_starts * info["sfreq"], gap_stops * info["sfreq"]
+                    )
```
Use int arithmetic where possible rather than relying on `np.arange` over float with a step of `1.`, just figure out how many elements there should be
```suggestion
                    int(round(stop * info["sfreq"])) - int(round(start * info["sfreq"]))
                    for start, stop in zip(gap_starts, gap_stops)
```

KristijanArmeni replied:
added in https://github.com/mne-tools/mne-python/pull/12279/commits/2c0861c03ecd79b1e65115440318929c4f3acdb3

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,31 +139,133 @@ def __init__(
+
+        seg_gap_dict = {}
+
+        if has_gaps:
```
Rather than three separate conditional `if has_gaps`, I wonder if we can get away with a single code path. In other words, make things a *tiny* bit less efficient by always doing the `gap_starts` etc. below and never have a conditional an `has_gaps`, and always constructing `Annotations`. And in the case where there are no gaps, these should become no-ops anyway, i.e., the `Annotations` will be empty (which is fine).

Usually code can be constructed this way and 1) it's a nice sanity check that it's coded properly, and 2) fewer branches mean less stuff to understand / think about when debugging. But if you try to implement things this way and it becomes a pain then I can live with the current branching!

KristijanArmeni replied:
Ok, got it! Will give it a shot! Thanks.

KristijanArmeni replied:
It seems branches were indeed unnecessary. Can make tests passing after removing the conditionals and adjusting accordingly in https://github.com/mne-tools/mne-python/pull/12279/commits/37197c7b18546bb77af7e075d691594d8ee1e854.

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,31 +139,133 @@ def __init__(
         """Read a chunk of raw data."""
+        from neo import Segment
         from neo.io import NeuralynxIO
+        from quantities import Hz
```
I had to look this up to make sure, so let's save the next dev the trouble
```suggestion
        # quantities is a dependency of neo so we are guaranteed it exists
        from quantities import Hz
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -188,15 +325,41 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        segments_arr = np.array(neo_block[0].segments, dtype=object)
+
+        # if gaps were detected correctly insert gap Segments in between valid Segments
+        if self._raw_extras[0]["seg_gap_dict"]:
```
... maybe my idea above about having a single code path won't work here. But if it does that's great!

larsoner reviewed (commented):
Looks good to me so far! One idea for a potential refactoring and a couple small cleanup ideas

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -81,12 +122,18 @@ def __init__(
-                + f"\nOriginal neo.NeuralynxIO.parse_header() ValueError:\n{e}"
-            )
+            # give a more informative error message and what the user can do about it
+            if "Incompatible section structures across streams" in e.__str__():
```
Better to use builtin public method than dunder
```suggestion
            if "Incompatible section structures across streams" in str(e):
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -81,12 +122,18 @@ def __init__(
+                    + f"\nOriginal neo.NeuralynxRawIO._parse_header() ValueError:\n{e}"
+                )
+            else:
+                raise e  # any other Neo error, just raise it
```
Traceback will be cleaner this way
```suggestion
                raise
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -81,12 +122,18 @@ def __init__(
+                    + "by excluding other channels with `exclude_fname_patterns` "
+                    + "input argument."
+                    + f"\nOriginal neo.NeuralynxRawIO._parse_header() ValueError:\n{e}"
+                )
```
You already give the error above so raise this one from None to avoid duplicate info
```suggestion
                ) from None
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -81,12 +122,18 @@ def __init__(
+                    + "Try reading in only channels with uniform sampling rate "
+                    + "by excluding other channels with `exclude_fname_patterns` "
+                    + "input argument."
+                    + f"\nOriginal neo.NeuralynxRawIO._parse_header() ValueError:\n{e}"
```
Better not to ref private function I think
```suggestion
                    + f"\nOriginal neo.NeuralynxRawIO ValueError:\n{e}"
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,32 +145,122 @@ def __init__(
-                for i in range(n_segments)
+                int(round(stop * info["sfreq"])) - int(round(start * info["sfreq"]))
+                for start, stop in zip(gap_starts, gap_stops)
             ]
+        ).astype(int)  # force an int array (if no gaps, empty array is a float)
```
Better way to accomplish the same thing
```suggestion
            ],
            dtype=int,
        )
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,32 +145,122 @@ def __init__(
+        all_starts_ids = np.argsort(np.concatenate([start_times, gap_starts]))
+
+        # variable indicating whether each segment is a gap or not
+        gap_indicator = np.concatenate(
+            [
+                np.full(len(start_times), fill_value=0),
+                np.full(len(gap_starts), fill_value=1),
+            ]
+        )
+        gap_indicator = gap_indicator[all_starts_ids].astype(bool)
```
```suggestion
        gap_indicator = np.concatenate(
            [
                np.zeros(len(start_times), bool),
                np.ones(len(gap_starts), bool),
            ]
        )
        gap_indicator = gap_indicator[all_starts_ids]
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,32 +145,122 @@ def __init__(
+        # store this in a dict to be passed to _raw_extras
+        seg_gap_dict = {
+            "gap_n_samps": gap_n_samps,
+            "isgap": gap_indicator,  # 0 (data segment) or 1 (gap segment)
```
```suggestion
            "isgap": gap_indicator,  # False (data segment) or True (gap segment)
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -98,32 +145,122 @@ def __init__(
+                mne_times[onset_id + (n - 1)] - mne_times[onset_id]
+                for onset_id, n in zip(gap_start_ids, gap_n_samps)
+            ],
+            description=["BAD_ACQ_SKIP" for _ in gap_start_ids],
```
```suggestion
            description=["BAD_ACQ_SKIP"] * len(gap_start_ids),
```

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -136,13 +273,7 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
-                for segment_id in range(len(neo_block[0].segments))
-            ]
-        )
+        segment_sizes = self._raw_extras[0]["segment_sizes"]
```
This will matter if you ever concatenate multiple RawNeuralynx instances
```suggestion
        segment_sizes = self._raw_extras[fi]["segment_sizes"]
```
Looks like we need a similar fix to `exclude_filename` above, maybe also other places

larsoner commented on mne/io/neuralynx/neuralynx.py:
```diff
@@ -188,15 +319,44 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        segments_arr = np.array(neo_block[0].segments, dtype=object)
+
+        # if gaps were detected, correctly insert gap Segments in between valid Segments
+        gap_samples = self._raw_extras[0]["seg_gap_dict"]["gap_n_samps"]
```
... like here
```suggestion
        gap_samples = self._raw_extras[fi]["seg_gap_dict"]["gap_n_samps"]
```

larsoner reviewed (commented):
Looks good! Just a bunch of tiny suggestions (and one fix).

Can you think of any way to get a small recording that has these gaps to add to the testing dataset?

KristijanArmeni commented:
> Looks good! Just a bunch of tiny suggestions (and one fix).

Thanks for feedback! Will look and sort out these shortly.

> Can you think of any way to get a small recording that has these gaps to add to the testing dataset?

I asked and these gaps occur quite randomly and my lab contact was not sure how to trigger those. An option I will look into is to read-in the intact recording with the Matlab reader shipped by Neuralynx, set some samples as invalid (i.e. cause a gap) and write back to .ncs with the Nlx2Mat.m reader that ships from Neuralynx.

I did find a [dataset with ring-buffer errors](https://gin.g-node.org/TRuikes/ephy_testing_data/src/master/neuralynx/examples_breaking_data/dsp_compensation_and_ringbuff_errors) via a an older git issue on `Neo` github. But I was not able to read it in locally with neo (chokes for an unrelated reason, seems the header does not have closing date info as the files were not closed properly or so)

larsoner commented:
Okay hopefully you can fake a file using the MATLAB software! If not then I think we can merge this as-is given you have been giving it real-world testing. And maybe someday you'll manage to collect some tiny dataset that has the issue.

KristijanArmeni commented:
> Okay hopefully you can fake a file using the MATLAB software! If not then I think we can merge this as-is given you have been giving it real-world testing. And maybe someday you'll manage to collect some tiny dataset that has the issue.

OK, I think it will work by artificially creating a gap via MATLAB! I can then save tweaked data back to `.ncs` and `.mat`. This allows us to set up multiple gap scenarios if need be. But it looks we're on good track (see a preliminary check below).

Not sure what a realistic test would like. Simplest would be to just test one or two such gaps artificially set gaps on one channel. And then see if our reader correctly infers these samples. The assumption would be that such gaps occur uniformly across channels.

I think I can prepare a testing script + testing dataset for this on Monday. Is the best practice to just write a new test script (e.g. `test_neuralynx_gap_detection.py` in addition to the existing `test_neuralynx.py`?

### Initial check
For a quick and dirty check, I tried setting 100 contiguous samples (50 msec data) in the middle of one of our testing `.ncs` channels to 0 and marking them as invalid (all via Matlab). I then save back to a) `modified_chan.ncs` and b) `modified_chan.mat`.

Reading in original the `modified_chan.mat` via `scipy.loadmat()` and the `modified_chan.ncs` via `read_raw_neuralynx()` (which infers missing samples), it passes `assert_allclose()`. So, looks pretty promising:
![comparison](https://github.com/mne-tools/mne-python/assets/14061041/23c9d621-c228-40df-b516-28d07ca08fc2)

larsoner commented:
> Is the best practice to just write a new test script (e.g. test_neuralynx_gap_detection.py in addition to the existing test_neuralynx.py?

If it were a big test sure. But I don't think this one would be hundreds of lines, so I'd just add it to `test_neuralynx.py` with a descriptive name.

> So, looks pretty promising:

👍

KristijanArmeni commented:
@larsoner updated `datasets/config.py` and added test in [e3d4d](https://github.com/mne-tools/mne-python/pull/12279/commits/e3d4d8ebf0e862a2c3173fb40b79c93a1ece595e). 

It seems CIs are failing due to some neo import issues (via `_soft_import` or so?). Looking at the error messages, can't really figure why -- is it anything related to the test script/read_raw_neuralynx? (it passes locally)

Otherwise, a recap: Created 3 gaps (130 invalid samples) in 2 .ncs files, then read them in with `read_raw_neuralynx`, check that it inferred the correct number of missing samples and test the loaded data against the "ground truth" files (.mat loaded in via scipy).

Also added a line to `devel.rst`.

larsoner commented on mne/io/neuralynx/tests/test_neuralynx.py:
```diff
@@ -136,5 +171,52 @@ def test_neuralynx():
+@requires_testing_data
+def test_neuralynx_gaps():
+    """Test gap detection."""
+    # ignore files with no gaps
```
This is one problem
```suggestion
    pytest.importorskip("neo")

    # ignore files with no gaps
```
but a worse problem is that it seems like `neo` is not being installed on any of those CIs. I'll push a commit

KristijanArmeni commented:
> Feel free to take out of draft mode if you're done then ping me @KristijanArmeni 

Thanks and done! @larsoner 

> and I'll merge assuming CIs are green (other than the Windows pip-pre one, which is having unrelated issues)

I pulled this latest branch locally and `test_neuralynx.py` passes fine. This is good for me as far as testing and implementation is concerned -- as far as I can understand the failing CIs are unrelated to this PR (?).

larsoner commented:
Some are, for example:
```
  └─ neo   does not exist (perhaps a typo or a missing channel).
```
is because *I changed environment.yml in this PR* to add neo, but used the wrong name. After I push a commit to fix this, all CIs should come back green other than windows pip-pre.

KristijanArmeni commented:
> is because _I changed environment.yml in this PR_ to add neo, but used the wrong name. After I push a commit to fix this, all CIs should come back green other than windows pip-pre.

Yeah, looks like the CIs are green, now @larsoner. Thanks! (This branch still needs a merge from main, but other than that it seems good)

larsoner commented:
Updated to latest `main` and marking for merge when green, thanks in advance @KristijanArmeni !
